File: bubus/event_history.py
from __future__ import annotations

import logging
from collections.abc import Awaitable, Callable
from datetime import UTC, datetime, timedelta
from typing import Any, Generic, Literal, TypeVar, overload

from .base_event import BaseEvent, UUIDStr

BaseEventT = TypeVar('BaseEventT', bound=BaseEvent[Any])
TExpectedEvent = TypeVar('TExpectedEvent', bound=BaseEvent[Any])
EventPatternType = str | Literal['*'] | type[BaseEvent[Any]]

logger = logging.getLogger('bubus')


class EventHistory(dict[UUIDStr, BaseEventT], Generic[BaseEventT]):
    """Ordered event history map with query and trim helpers."""

    __slots__ = ('max_history_size', 'max_history_drop', '_warned_about_dropping_uncompleted_events')

    def __init__(self, max_history_size: int | None = 100, max_history_drop: bool = False):
        super().__init__()
        self.max_history_size = max_history_size
        self.max_history_drop = max_history_drop
        self._warned_about_dropping_uncompleted_events = False

    def add_event(self, event: BaseEventT) -> None:
        self[event.event_id] = event

    def get_event(self, event_id: str) -> BaseEventT | None:
        return self.get(event_id)

    def remove_event(self, event_id: str) -> bool:
        if event_id not in self:
            return False
        del self[event_id]
        return True

    def has_event(self, event_id: str) -> bool:
        return event_id in self

    @staticmethod
    def normalize_event_pattern(event_pattern: EventPatternType) -> str:
        if event_pattern == '*':
            return '*'
        if isinstance(event_pattern, str):
            return event_pattern
        event_type_field = event_pattern.model_fields.get('event_type')
        event_type_default = event_type_field.default if event_type_field is not None else None
        if isinstance(event_type_default, str) and event_type_default not in ('', 'UndefinedEvent'):
            return event_type_default
        return event_pattern.__name__

    @staticmethod
    def is_event_complete_fast(event: BaseEvent[Any]) -> bool:
        signal = event._event_completed_signal  # pyright: ignore[reportPrivateUsage]
        if signal is not None:
            return signal.is_set()
        if event._event_is_complete_flag:  # pyright: ignore[reportPrivateUsage]
            return True
        return event.event_completed_at is not None

    def event_is_child_of(self, event: BaseEvent[Any], ancestor: BaseEvent[Any]) -> bool:
        current_id = event.event_parent_id
        visited: set[str] = set()

        while current_id and current_id not in visited:
            if current_id == ancestor.event_id:
                return True
            visited.add(current_id)
            parent = self.get(current_id)
            if parent is None:
                return False
            current_id = parent.event_parent_id

        return False

    @overload
    async def find(
        self,
        event_type: type[TExpectedEvent],
        where: None = None,
        child_of: BaseEvent[Any] | None = None,
        past: bool | float | timedelta | None = None,
        future: bool | float | None = None,
        event_is_child_of: Callable[[BaseEvent[Any], BaseEvent[Any]], bool] | None = None,
        wait_for_future_match: Callable[
            [str, Callable[[BaseEvent[Any]], bool], bool | float],
            Awaitable[BaseEvent[Any] | None],
        ]
        | None = None,
        **event_fields: Any,
    ) -> TExpectedEvent | None: ...

    @overload
    async def find(
        self,
        event_type: type[TExpectedEvent],
        where: Callable[[TExpectedEvent], bool],
        child_of: BaseEvent[Any] | None = None,
        past: bool | float | timedelta | None = None,
        future: bool | float | None = None,
        event_is_child_of: Callable[[BaseEvent[Any], BaseEvent[Any]], bool] | None = None,
        wait_for_future_match: Callable[
            [str, Callable[[BaseEvent[Any]], bool], bool | float],
            Awaitable[BaseEvent[Any] | None],
        ]
        | None = None,
        **event_fields: Any,
    ) -> TExpectedEvent | None: ...

    @overload
    async def find(
        self,
        event_type: str | Literal['*'],
        where: Callable[[BaseEvent[Any]], bool] | None = None,
        child_of: BaseEvent[Any] | None = None,
        past: bool | float | timedelta | None = None,
        future: bool | float | None = None,
        event_is_child_of: Callable[[BaseEvent[Any], BaseEvent[Any]], bool] | None = None,
        wait_for_future_match: Callable[
            [str, Callable[[BaseEvent[Any]], bool], bool | float],
            Awaitable[BaseEvent[Any] | None],
        ]
        | None = None,
        **event_fields: Any,
    ) -> BaseEvent[Any] | None: ...

    async def find(
        self,
        event_type: EventPatternType,
        where: Callable[[Any], bool] | None = None,
        child_of: BaseEvent[Any] | None = None,
        past: bool | float | timedelta | None = None,
        future: bool | float | None = None,
        event_is_child_of: Callable[[BaseEvent[Any], BaseEvent[Any]], bool] | None = None,
        wait_for_future_match: Callable[
            [str, Callable[[BaseEvent[Any]], bool], bool | float],
            Awaitable[BaseEvent[Any] | None],
        ]
        | None = None,
        **event_fields: Any,
    ) -> BaseEvent[Any] | None:
        resolved_past_input = True if past is None else past
        if isinstance(resolved_past_input, timedelta):
            resolved_past: bool | float = max(0.0, resolved_past_input.total_seconds())
        elif isinstance(resolved_past_input, bool):
            resolved_past = resolved_past_input
        else:
            resolved_past = max(0.0, float(resolved_past_input))

        resolved_future_input = False if future is None else future
        if isinstance(resolved_future_input, bool):
            resolved_future: bool | float = resolved_future_input
        else:
            resolved_future = max(0.0, float(resolved_future_input))

        if resolved_past is False and resolved_future is False:
            return None

        event_key = self.normalize_event_pattern(event_type)
        where_predicate: Callable[[BaseEvent[Any]], bool]
        if where is None:
            where_predicate = lambda _: True
        else:
            where_predicate = where

        child_check = event_is_child_of or self.event_is_child_of
        cutoff: datetime | None = None
        if resolved_past is not True:
            cutoff = datetime.now(UTC) - timedelta(seconds=float(resolved_past))
        missing = object()

        def matches(event: BaseEvent[Any]) -> bool:
            if event_key != '*' and event.event_type != event_key:
                return False
            if child_of is not None and not child_check(event, child_of):
                return False
            field_mismatch = any(
                getattr(event, field_name, missing) != expected_value
                for field_name, expected_value in event_fields.items()
            )
            if field_mismatch:
                return False
            if not where_predicate(event):
                return False
            return True

        if resolved_past is not False:
            events = list(self.values())
            for event in reversed(events):
                if cutoff is not None and event.event_created_at < cutoff:
                    continue
                if matches(event):
                    return event

        if resolved_future is False or wait_for_future_match is None:
            return None

        return await wait_for_future_match(event_key, matches, resolved_future)

    def cleanup_excess_events(self, *, on_remove: Callable[[BaseEventT], None] | None = None) -> int:
        if self.max_history_size is None:
            return 0
        if self.max_history_size == 0:
            return self.trim_event_history(on_remove=on_remove)
        if len(self) <= self.max_history_size:
            return 0

        total_events = len(self)
        remove_count = total_events - self.max_history_size
        event_ids_to_remove = list(self.keys())[:remove_count]

        removed_count = 0
        for event_id in event_ids_to_remove:
            event = self.get(event_id)
            if event is None:
                continue
            del self[event_id]
            if on_remove is not None:
                on_remove(event)
            removed_count += 1

        return removed_count

    def trim_event_history(
        self,
        *,
        on_remove: Callable[[BaseEventT], None] | None = None,
        owner_label: str | None = None,
    ) -> int:
        if self.max_history_size is None:
            return 0

        if self.max_history_size == 0:
            completed_event_ids = [event_id for event_id, event in self.items() if self.is_event_complete_fast(event)]
            removed_count = 0
            for event_id in completed_event_ids:
                event = self.get(event_id)
                if event is None:
                    continue
                del self[event_id]
                if on_remove is not None:
                    on_remove(event)
                removed_count += 1
            return removed_count

        if not self.max_history_drop:
            return 0

        if len(self) <= self.max_history_size:
            return 0

        remaining_overage = len(self) - self.max_history_size
        removed_count = 0

        for event_id, event in list(self.items()):
            if remaining_overage <= 0:
                break
            if not self.is_event_complete_fast(event):
                continue
            del self[event_id]
            if on_remove is not None:
                on_remove(event)
            removed_count += 1
            remaining_overage -= 1

        dropped_uncompleted = 0
        if remaining_overage > 0:
            for event_id, event in list(self.items()):
                if remaining_overage <= 0:
                    break
                if not self.is_event_complete_fast(event):
                    dropped_uncompleted += 1
                del self[event_id]
                if on_remove is not None:
                    on_remove(event)
                removed_count += 1
                remaining_overage -= 1

        if dropped_uncompleted > 0 and not self._warned_about_dropping_uncompleted_events:
            self._warned_about_dropping_uncompleted_events = True
            owner = owner_label or 'EventBus'
            logger.warning(
                '[bubus] ⚠️ Bus %s has exceeded max_history_size=%s and is dropping oldest history entries '
                '(even uncompleted events). Increase max_history_size or set max_history_drop=False to reject.',
                owner,
                self.max_history_size,
            )

        return removed_count

File: bubus-ts/src/event_history.ts
import { BaseEvent } from './base_event.js'
import type { EventPattern, FindWindow } from './types.js'
import { normalizeEventPattern } from './types.js'
import { monotonicDatetime } from './helpers.js'

export type EventHistoryFindOptions = {
  past?: FindWindow
  future?: FindWindow
  child_of?: BaseEvent | null
  event_is_child_of?: (event: BaseEvent, ancestor: BaseEvent) => boolean
  wait_for_future_match?: (
    event_pattern: string | '*',
    matches: (event: BaseEvent) => boolean,
    future: FindWindow
  ) => Promise<BaseEvent | null>
} & Record<string, unknown>

export type EventHistoryTrimOptions<TEvent extends BaseEvent = BaseEvent> = {
  is_event_complete?: (event: TEvent) => boolean
  on_remove?: (event: TEvent) => void
  owner_label?: string
  max_history_size?: number | null
  max_history_drop?: boolean
}

export class EventHistory<TEvent extends BaseEvent = BaseEvent> implements Iterable<[string, TEvent]> {
  max_history_size: number | null
  max_history_drop: boolean

  private _events: Map<string, TEvent>
  private _warned_about_dropping_uncompleted_events: boolean

  constructor(options: { max_history_size?: number | null; max_history_drop?: boolean } = {}) {
    this.max_history_size = options.max_history_size === undefined ? 100 : options.max_history_size
    this.max_history_drop = options.max_history_drop ?? false
    this._events = new Map()
    this._warned_about_dropping_uncompleted_events = false
  }

  get size(): number {
    return this._events.size
  }

  [Symbol.iterator](): Iterator<[string, TEvent]> {
    return this._events[Symbol.iterator]()
  }

  entries(): IterableIterator<[string, TEvent]> {
    return this._events.entries()
  }

  keys(): IterableIterator<string> {
    return this._events.keys()
  }

  values(): IterableIterator<TEvent> {
    return this._events.values()
  }

  clear(): void {
    this._events.clear()
  }

  get(event_id: string): TEvent | undefined {
    return this._events.get(event_id)
  }

  set(event_id: string, event: TEvent): this {
    this._events.set(event_id, event)
    return this
  }

  has(event_id: string): boolean {
    return this._events.has(event_id)
  }

  delete(event_id: string): boolean {
    return this._events.delete(event_id)
  }

  addEvent(event: TEvent): void {
    this._events.set(event.event_id, event)
  }

  getEvent(event_id: string): TEvent | undefined {
    return this._events.get(event_id)
  }

  removeEvent(event_id: string): boolean {
    return this._events.delete(event_id)
  }

  hasEvent(event_id: string): boolean {
    return this._events.has(event_id)
  }

  static normalizeEventPattern(event_pattern: EventPattern | '*'): string | '*' {
    return normalizeEventPattern(event_pattern)
  }

  find(event_pattern: '*', where?: (event: TEvent) => boolean, options?: EventHistoryFindOptions): Promise<TEvent | null>
  find<TMatch extends TEvent>(
    event_pattern: EventPattern<TMatch>,
    where?: (event: TMatch) => boolean,
    options?: EventHistoryFindOptions
  ): Promise<TMatch | null>
  async find(
    event_pattern: EventPattern<TEvent> | '*',
    where: (event: TEvent) => boolean = () => true,
    options: EventHistoryFindOptions = {}
  ): Promise<TEvent | null> {
    const past = options.past ?? true
    const future = options.future ?? false
    const child_of = options.child_of ?? null
    const eventIsChildOf = options.event_is_child_of ?? ((event: BaseEvent, ancestor: BaseEvent) => this.eventIsChildOf(event, ancestor))
    const waitForFutureMatch = options.wait_for_future_match
    if (past === false && future === false) {
      return null
    }

    const event_key = EventHistory.normalizeEventPattern(event_pattern)
    const cutoff_at = past === true ? null : monotonicDatetime(new Date(Date.now() - Math.max(0, Number(past)) * 1000).toISOString())

    const event_field_filters = Object.entries(options).filter(
      ([key, value]) =>
        key !== 'past' &&
        key !== 'future' &&
        key !== 'child_of' &&
        key !== 'event_is_child_of' &&
        key !== 'wait_for_future_match' &&
        value !== undefined
    )

    const matches = (event: BaseEvent): boolean =>
      (event_key === '*' || event.event_type === event_key) &&
      (!child_of || eventIsChildOf(event, child_of)) &&
      event_field_filters.every(([field_name, expected]) => (event as unknown as Record<string, unknown>)[field_name] === expected) &&
      where(event as TEvent)

    if (past !== false) {
      const history_values = Array.from(this._events.values())
      for (let i = history_values.length - 1; i >= 0; i -= 1) {
        const event = history_values[i]
        if (cutoff_at !== null && event.event_created_at < cutoff_at) {
          continue
        }
        if (matches(event)) {
          return event
        }
      }
    }

    if (future === false || !waitForFutureMatch) {
      return null
    }

    return (await waitForFutureMatch(event_key, matches, future)) as TEvent | null
  }

  cleanupExcessEvents(options: EventHistoryTrimOptions<TEvent> = {}): number {
    const max_history_size = options.max_history_size ?? this.max_history_size
    if (max_history_size === null) {
      return 0
    }
    if (max_history_size === 0) {
      return this.trimEventHistory(options)
    }
    const remove_count = this.size - max_history_size
    if (remove_count <= 0) {
      return 0
    }

    const on_remove = options.on_remove
    let removed_count = 0

    for (const event_id of Array.from(this._events.keys()).slice(0, remove_count)) {
      const event = this._events.get(event_id)
      if (!event) {
        continue
      }
      this._events.delete(event_id)
      on_remove?.(event)
      removed_count += 1
    }

    return removed_count
  }

  trimEventHistory(options: EventHistoryTrimOptions<TEvent> = {}): number {
    const max_history_size = options.max_history_size ?? this.max_history_size
    const max_history_drop = options.max_history_drop ?? this.max_history_drop
    if (max_history_size === null) {
      return 0
    }

    const is_event_complete = options.is_event_complete ?? ((event: TEvent) => event.event_status === 'completed')
    const on_remove = options.on_remove

    if (max_history_size === 0) {
      let removed_count = 0
      for (const [event_id, event] of Array.from(this._events.entries())) {
        if (!is_event_complete(event)) {
          continue
        }
        this._events.delete(event_id)
        on_remove?.(event)
        removed_count += 1
      }
      return removed_count
    }

    if (!max_history_drop || this.size <= max_history_size) {
      return 0
    }

    let remaining_overage = this.size - max_history_size
    let removed_count = 0
    const remove_event = (event_id: string, event: TEvent): void => {
      this._events.delete(event_id)
      on_remove?.(event)
      removed_count += 1
    }

    for (const [event_id, event] of Array.from(this._events.entries())) {
      if (remaining_overage <= 0) {
        break
      }
      if (!is_event_complete(event)) {
        continue
      }
      remove_event(event_id, event)
      remaining_overage -= 1
    }

    let dropped_uncompleted = 0
    for (const [event_id, event] of Array.from(this._events.entries())) {
      if (remaining_overage <= 0) {
        break
      }
      if (!is_event_complete(event)) {
        dropped_uncompleted += 1
      }
      remove_event(event_id, event)
      remaining_overage -= 1
    }

    if (dropped_uncompleted > 0 && !this._warned_about_dropping_uncompleted_events) {
      this._warned_about_dropping_uncompleted_events = true
      const owner_label = options.owner_label ?? 'EventBus'
      console.error(
        `[bubus] ⚠️ Bus ${owner_label} has exceeded max_history_size=${max_history_size} and is dropping oldest history entries (even uncompleted events). Increase max_history_size or set max_history_drop=false to reject.`
      )
    }

    return removed_count
  }

  private eventIsChildOf(event: BaseEvent, ancestor: BaseEvent): boolean {
    let current_parent_id = event.event_parent_id
    const visited = new Set<string>()

    while (current_parent_id && !visited.has(current_parent_id)) {
      if (current_parent_id === ancestor.event_id) {
        return true
      }
      visited.add(current_parent_id)
      const parent = this._events.get(current_parent_id)
      if (!parent) {
        return false
      }
      current_parent_id = parent.event_parent_id
    }

    return false
  }
}
